### ISSUE TITLE
debug mode is utterly broken

### ISSUE THREAD

**Issue by avikivity** [scylladb/scylladb#14196]
A simple `gdb build/debug/scylla -ex start` shows all variables as optimized out, making it impossible to debug.

**Comment by avikivity** [comment]
Bisected to 7e68ed6a5de2635dab3e7169a3cc9c2d79e29a88..

**Comment by kostja** [comment]
Thanks so much for this. It never worked for me and I didn't complain because I thought it's due to a non-standard build toolchain.

**Comment by nyh** [comment]
> /cc @michoecho @nyh re people don't complain

I agree with this statement, but understand the situation a bit different from you:

When we were developing OSv, we also had a "release" and a "debug" build, and those were exactly what the names suggested - "release" was performance-oriented, and "debug" was used when you wanted to use the debugger (gdb) - so it had a lower optimization.

But when we moved to Scylla, at some point you decided that the "debug" build should also have sanitizers enabled. This made the executables huge if not immediately stripped (https://github.com/scylladb/scylladb/issues/5), and super-slow, sometimes as much as 100 times slower than the release build, which made it a huge pain to use gdb on these executables, so nobody did. And nobody complained because everybody thought (rightly) that this was deliberate: that our "debug" builds aren't for "debugging" (with gdb) but like "#ifdef DEBUG" simply have much more internal testing. When two years ago someone broke the debugging ability on these builds, nobody cared (it's now clear that @kostja noticed, but he wasn't very surprised that it doesn't work). When people rarely do use gdb on Scylla, it's usually with a customer core dump and the executable is the unstripped release build - *not* a "debug" build.

I honestly think the real solution is to separate the "gdb" and "sanitizer + extra-checks" build modes. Anything else will mean that either one or the other isn't usable for ordinary users, but nobody will complain because they rightly believe that this was done deliberately and it's pointless to complain.

**Comment by avikivity** [comment]
Fix was reverted in b7627085cb13449bfcabf303f486bfd88bf4526c. @michoecho claims the revert was unnecessary, but I'd like to measure before trying again.

**Comment by avikivity** [comment]
Removing backport candidate label as the issue was reopened.

**Comment by nyh** [comment]
> Fix was reverted in [b762708](https://github.com/scylladb/scylladb/commit/b7627085cb13449bfcabf303f486bfd88bf4526c). @michoecho claims the revert was unnecessary, but I'd like to measure before trying again.

IIRC, the estimate was that the completely-unoptimized tests on Jenkins are **twice** slower than the ones we have before this patch. Why is this acceptable?

**Comment by avikivity** [comment]
Twice as slow is too slow. I'd accept a 30% regression in ./test.py --mode debug (which would be less for the entire pipeline).

But we need to measure, not estimate.

**Comment by mykaul** [comment]
Is this still relevant? Is it still tracking the same issue as originally reported?

**Comment by mykaul** [comment]
ping @avikivity ?

**Comment by mykaul** [comment]
re-ping @avikivity - relevant for 5.4, or can be deferred?

**Comment by avikivity** [comment]
It's still open. It's not necessary for any release as it's only about developer comfort when using debug mode.

There's a chance it was fixed or will be fixed by a toolchain update, need to check.

**Comment by avikivity** [comment]
Just checked, still relevant.

**Comment by alezzqz** [comment]
> /cc @michoecho @nyh re people don't complain

people don't complain because they use
`tools/toolchain/dbuild ./configure.py --mode=debug --cflags=-Og`
and think this bevavior is normal )
